crates/network/src/net_thread.rs

```diff
@@ -0,0 +1,175 @@
+use std::net::SocketAddr;
+
+use rjacraft_protocol::{frame::*, packets::*, ProtocolType, ProtocolVersion};
+use tokio::{io, net, task};
+use tracing::*;
+
+// Some of these are invalid at some states
+pub enum PeerMsgIn {
+    Drop,
+    StatusPacket(cb::StatusPacket),
+    // LoginPacket(cb::LoginPacket),
+    ConfigurationPacket(cb::ConfigurationPacket),
+    PlayPacket(cb::PlayPacket),
+}
+
+pub enum PeerMsgOut {
+    Disconnected,
+    HandshakeComplete(ProtocolVersion, String, u16),
+    NeedStatus,
+    // Authenticate(String, Option<uuid::Uuid>),
+    // NeedConfiguration,
+    // PlayPacket(sb::PlayPacket),
+}
+
+#[derive(Debug, Clone, Copy)]
+enum ConnectionState {
+    Handshake,
+    Status,
+    Login(ProtocolVersion),
+    // Configuration,
+    // Play,
+}
+
+#[derive(Debug, thiserror::Error)]
+enum PeerLoopError {
+    #[error(transparent)]
+    Reading(#[from] ReadPacketError),
+    #[error(transparent)]
+    Writing(#[from] WritePacketError),
+    #[error(transparent)]
+    DecodingHandshake(#[from] sb::HandshakePacketDecodeError),
+    #[error(transparent)]
+    DecodingStatus(#[from] sb::StatusPacketDecodeError),
+    #[error(transparent)]
+    EncodingStatus(#[from] cb::StatusPacketEncodeError),
+    #[error("Wrong protocol version: {0:?}")]
+    WrongVersion(rjacraft_protocol::ProtocolVersion),
+}
+
+async fn peer_loop(
+    stream: net::TcpStream,
+    msg_in: flume::Receiver<PeerMsgIn>,
+    msg_out: flume::Sender<PeerMsgOut>,
+) -> Result<(), PeerLoopError> {
+    let (mut read, mut write) = stream.into_split();
+
+    // reading varints isn't cancellable so here we are
+    let (frames_tx, frames_rx) = flume::unbounded();
+    let mut read_task: task::JoinHandle<Result<(), ReadPacketError>> = tokio::spawn(async move {
+        loop {
+            let packet = read_frame(&mut read).await?;
+
+            frames_tx.send(packet).unwrap();
```

Comment:
Please add an explicit doc on why the panic is the best approach here / not possible under normal circumstances.

Reply:
Because message channels don't really crash and that error is mostly pointless